src/storage/sql_table.cpp

```diff
@@ -86,6 +66,59 @@ ProjectionMap SqlTable::ProjectionMapForOids(const std::vector<catalog::col_oid_
   return projection_map;
 }
 
+void SqlTable::Reset() { table_.data_table_->Reset(); }
+
+void SqlTable::CopyTable(const common::ManagedPointer<transaction::TransactionContext> txn,
+                         const common::ManagedPointer<SqlTable> src) {
+  auto it = src->begin();
+  std::vector<catalog::col_oid_t> col_oids;
+  for (auto &cols : table_.column_map_) {
+    col_oids.push_back(cols.first);
+  }
+  auto pr_init = InitializerForProjectedRow(col_oids);
+  // Allocate on stack since we would otherwise free it immediately and lifecycle is short
+  void *buffer = alloca(pr_init.ProjectedRowSize());
+  auto *projected_row = pr_init.InitializeRow(buffer);
+  while (it != src->end()) {
+    const TupleSlot slot = *it;
+    // Only fill the buffer with valid, visible tuples
+    if (!src->Select(txn, slot, projected_row)) {
+      it++;
+      continue;
+    }
+
+    // TODO(tanujnay112) I don't like how I have to hardcode this
+    auto *redo = txn->StageWrite(
+        catalog::MakeTempOid<catalog::db_oid_t>(catalog::INVALID_DATABASE_OID.UnderlyingValue()),
+        catalog::MakeTempOid<catalog::table_oid_t>(catalog::INVALID_TABLE_OID.UnderlyingValue()), pr_init);
+    auto *new_pr = redo->Delta();
+    auto pr_map = ProjectionMapForOids(col_oids);
+    for (auto &cols : table_.column_map_) {
+      auto offset = pr_map[cols.first];
+      auto new_pr_ptr = new_pr->AccessForceNotNull(offset);
+      auto src_ptr = projected_row->AccessWithNullCheck(offset);
+      if (src_ptr == nullptr) {
+        new_pr->SetNull(offset);
+        continue;
+      }
+      std::memcpy(new_pr_ptr, src_ptr, table_.layout_.AttrSize(cols.second.col_id_));
+
+      // copy over varlens contents
+      if (table_.layout_.IsVarlen(cols.second.col_id_)) {
+        auto varlen = reinterpret_cast<storage::VarlenEntry *>(src_ptr);
+        if (varlen->NeedReclaim()) {
+          byte *new_allocation = new byte[varlen->Size()];
```

Comment:
I don't think that's how we usually allocate the extra buffer for `VarlenEntry`. I think it's likely going through `AllocateAligned` but we should double check.

Reply:
Agreed. It looks like `AllocateAligned` is the way we typically do this in those cases where the `VarlenEntry` is not backed by a string of some sort. I've updated the implementation.